File: SonnetDB/web/src/api/events.ts
/**
 * SSE 客户端封装：基于浏览器原生 EventSource，订阅服务端 `/v1/events`。
 *
 * - token 通过 query string `?access_token=` 传递（EventSource 不支持自定义 header）。
 * - 失败时由浏览器内置的指数退避重连机制处理；本封装额外处理 401 — 直接关闭并通知调用方。
 * - channels 参数转换为 `?stream=metrics,slow_query,db`，缺省订阅全部。
 */
export type ServerEventType = 'metrics' | 'slow_query' | 'db' | 'hello';

export interface ServerEventPayload {
  type: ServerEventType;
  /** 服务端事件产生时间（Unix ms）。来自 SSE id 字段，hello 事件可能为 0。 */
  timestampMs: number;
  /** 已解析的 JSON 数据。 */
  data: unknown;
}

export interface MetricsSnapshot {
  uptimeSeconds: number;
  databases: number;
  sqlRequests: number;
  sqlErrors: number;
  rowsInserted: number;
  rowsReturned: number;
  subscriberCount: number;
  perDatabaseSegments: Record<string, number>;
}

export interface SlowQueryPayload {
  database: string;
  sql: string;
  elapsedMs: number;
  rowCount: number;
  recordsAffected: number;
  failed: boolean;
}

export interface DatabasePayload {
  database: string;
  /** "created" / "dropped" */
  action: string;
}

export interface SseSubscribeOptions {
  channels?: ServerEventType[];
  onEvent: (evt: ServerEventPayload) => void;
  onError?: (status: 'error' | 'unauthorized') => void;
  onOpen?: () => void;
}

/**
 * 订阅 SSE 流，返回一个关闭函数。
 */
export function subscribeServerEvents(token: string, opts: SseSubscribeOptions): () => void {
  const params = new URLSearchParams();
  params.set('access_token', token);
  if (opts.channels && opts.channels.length > 0) {
    params.set('stream', opts.channels.join(','));
  }
  const url = `/v1/events?${params.toString()}`;
  const es = new EventSource(url);

  const handlers: Array<[string, (e: MessageEvent) => void]> = [];
  function on(type: string): void {
    const h = (e: MessageEvent) => {
      let data: unknown = null;
      try { data = JSON.parse(e.data); } catch { data = e.data; }
      const id = Number(e.lastEventId);
      opts.onEvent({
        type: type as ServerEventType,
        timestampMs: Number.isFinite(id) ? id : 0,
        data,
      });
    };
    es.addEventListener(type, h as EventListener);
    handlers.push([type, h]);
  }
  on('hello');
  on('metrics');
  on('slow_query');
  on('db');

  es.onopen = () => opts.onOpen?.();
  es.onerror = () => {
    // EventSource 在 401/网络错误都会触发 onerror；readyState=CLOSED 表示无法恢复
    if (es.readyState === EventSource.CLOSED) {
      opts.onError?.('unauthorized');
    } else {
      opts.onError?.('error');
    }
  };

  return () => {
    for (const [type, h] of handlers) {
      es.removeEventListener(type, h as EventListener);
    }
    es.close();
  };
}

File: SonnetDB/src/SonnetDB/Contracts/Events.cs
namespace SonnetDB.Contracts;

/// <summary>
/// 推送给 SSE 订阅者的统一事件包装。<see cref="Type"/> 决定 <see cref="Data"/> 的具体形状。
/// </summary>
/// <param name="Type">事件类型：<c>metrics</c> / <c>slow_query</c> / <c>db</c>。</param>
/// <param name="Data">JSON 序列化后的事件数据。</param>
/// <param name="TimestampMs">事件产生时间（Unix 毫秒，UTC）。</param>
public sealed record ServerEvent(string Type, string Data, long TimestampMs)
{
    /// <summary>事件 channel：周期性指标快照。</summary>
    public const string ChannelMetrics = "metrics";

    /// <summary>事件 channel：慢查询。</summary>
    public const string ChannelSlowQuery = "slow_query";

    /// <summary>事件 channel：数据库 DDL（CREATE/DROP）。</summary>
    public const string ChannelDatabase = "db";
}

/// <summary>
/// 周期性指标快照（与 Prometheus 文本相同口径，但 JSON 友好，便于前端图表/卡片直接绑定）。
/// </summary>
/// <param name="UptimeSeconds">服务运行秒数。</param>
/// <param name="Databases">已注册的数据库数量。</param>
/// <param name="SqlRequests">累计 SQL 请求数。</param>
/// <param name="SqlErrors">累计 SQL 错误数。</param>
/// <param name="RowsInserted">累计 INSERT 行数。</param>
/// <param name="RowsReturned">累计 SELECT 返回行数。</param>
/// <param name="SubscriberCount">当前 SSE 订阅者数量。</param>
/// <param name="PerDatabaseSegments">各数据库当前活跃 segment 数。</param>
public sealed record MetricsSnapshotEvent(
    double UptimeSeconds,
    int Databases,
    long SqlRequests,
    long SqlErrors,
    long RowsInserted,
    long RowsReturned,
    int SubscriberCount,
    IReadOnlyDictionary<string, int> PerDatabaseSegments);

/// <summary>
/// 慢查询事件：单条 SQL 执行耗时超过阈值时上报。
/// </summary>
/// <param name="Database">数据库名（控制面 SQL 时为 <c>"__control"</c>）。</param>
/// <param name="Sql">原始 SQL 文本（已截断到最多 1024 字符）。</param>
/// <param name="ElapsedMs">服务端执行耗时（毫秒）。</param>
/// <param name="RowCount">返回行数（非 SELECT 为 0）。</param>
/// <param name="RecordsAffected">受影响行数（非 SELECT 时有效；SELECT 为 -1）。</param>
/// <param name="Failed">是否因异常失败。</param>
/// <param name="Severity">慢查询等级：<c>slow</c> / <c>warning</c> / <c>critical</c>。</param>
public sealed record SlowQueryEvent(
    string Database,
    string Sql,
    double ElapsedMs,
    long RowCount,
    int RecordsAffected,
    bool Failed,
    string Severity);

/// <summary>
/// 慢查询事件严重等级。
/// </summary>
public static class SlowQuerySeverity
{
    /// <summary>达到基础慢查询阈值。</summary>
    public const string Slow = "slow";

    /// <summary>达到警告级阈值。</summary>
    public const string Warning = "warning";

    /// <summary>达到严重级阈值。</summary>
    public const string Critical = "critical";
}

/// <summary>
/// 数据库生命周期事件（CREATE / DROP）。
/// </summary>
/// <param name="Database">数据库名。</param>
/// <param name="Action">操作：<c>created</c> / <c>dropped</c>。</param>
public sealed record DatabaseEvent(string Database, string Action)
{
    /// <summary>CREATE DATABASE 后广播。</summary>
    public const string ActionCreated = "created";

    /// <summary>DROP DATABASE 后广播。</summary>
    public const string ActionDropped = "dropped";
}
